File: benchmarks/families/lstm.py
#!/usr/bin/env python3
"""The LSTM family in PyTorch, defined once, driven four ways.

PLAN.md; the counterpart of lstm.cpp. LSTM forecasting on UCI Beijing PM2.5,
predicting the next hourly reading from a window of past ones.

  lstm.py train    <csv> <csv> [epochs] [batch,...] [hidden] [past] [dev] [prec]
  lstm.py infer    <csv>       [reps]   [batch,...] [hidden] [past] [dev] [prec]
  lstm.py capacity <csv>       [batch]              [hidden] [past] [dev] [prec]

`nn.LSTM` reaches cuDNN's fused RNN, and so does OpenNN's
`LongShortTermMemoryOperator`. Both engines therefore run the *same NVIDIA
kernel*, which makes this the cleanest cell in the matrix: the arithmetic is
identical, so what is measured is the surrounding machinery -- data movement,
launch overhead, the optimiser -- rather than two hand-written kernels.

`nn.LSTM(features, hidden) + nn.Linear(hidden, 1)` is what lstm.cpp's
ForecastingLstmNetwork builds, and the parameter counts are printed by both so
the claim is checkable rather than asserted: 4*(hidden*features + hidden^2 +
2*hidden) + hidden + 1, which is 73,857 at the defaults.
"""

from __future__ import annotations

import contextlib
import csv
import os
import re
import sys
import time
from pathlib import Path

import numpy as np
import torch
from torch import nn

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

SEED = 42

class Forecaster(nn.Module):
    def __init__(self, features: int, hidden: int):
        super().__init__()
        self.lstm = nn.LSTM(features, hidden, batch_first=True)
        self.head = nn.Linear(hidden, 1)

    def forward(self, x):
        output, _ = self.lstm(x)
        return self.head(output[:, -1, :])

def report_blas() -> None:
    """Which BLAS this engine dispatches to, printed like OpenNN prints it.

    PyTorch's is fixed at build time -- the wheels are MKL-backed -- while
    OpenNN's is chosen at runtime. Recording both is what stops a comparison
    quietly turning into Eigen against MKL.
    """
    settings = torch.__config__.show()
    match = re.search(r"BLAS_INFO=(\w+)", settings)

    print(f"blas={match.group(1) if match else 'unknown'}")


def build(features: int, opts: dict) -> nn.Module:
    """The LSTM family. Nothing else here constructs the network."""
    torch.manual_seed(SEED)
    return Forecaster(features, opts["hidden"]).to(opts["device"])

def load_series(path: str, past: int) -> tuple[np.ndarray, np.ndarray]:
    """Windows of `past` hourly rows, predicting the next target.

    The last column is the target, matching what the tabular loader on the
    OpenNN side treats as such, and it is standardised on training statistics
    so neither engine pays a scaling stage the other does not.
    """
    report_opened(path)

    with open(path, newline="") as handle:
        rows = list(csv.reader(handle))

    values = np.asarray(rows[1:], dtype=np.float32)
    features, target = values[:, :-1], values[:, -1:]

    mean, std = features.mean(axis=0), features.std(axis=0)
    features = (features - mean) / np.where(std > 1.0e-12, std, 1.0)

    count = len(values) - past
    windows = np.lib.stride_tricks.sliding_window_view(features, past, axis=0)
    windows = np.ascontiguousarray(windows[:count].transpose(0, 2, 1))

    return windows, np.ascontiguousarray(target[past:past + count])

def resident_mib() -> float:
    """Resident set, MiB. The framework baseline is subtracted from the peak,
    because torch's import alone is ~816 MiB here against OpenNN's 209 -- so a
    raw RSS comparison measures which framework is bigger, not which run costs
    more, and the answer flips with dataset size."""
    try:
        with open("/proc/self/statm") as handle:
            return int(handle.read().split()[1]) * os.sysconf("SC_PAGE_SIZE") / (1024.0 * 1024.0)
    except Exception:
        return 0.0

def report_opened(path: str) -> None:
    """Announce the file actually opened, not the one passed in."""
    print(f"dataset_opened={Path(path).resolve()}", flush=True)

def parse_opts(argv: list[str], first: int) -> dict:
    def at(index: int, default: str) -> str:
        return argv[index] if len(argv) > index else default

    device = at(first + 2, "cuda")
    precision = at(first + 3, "fp32")

    if device == "cuda" and not torch.cuda.is_available():
        raise SystemExit("cuda requested but not available")

    # The runner pins CPU work to performance cores and sets a matched thread
    # count for both engines; honour it, or this side quietly takes every
    # logical CPU and the comparison becomes a thread-count comparison.
    threads = os.environ.get("TORCH_NUM_THREADS") or os.environ.get("OMP_NUM_THREADS")
    if threads:
        torch.set_num_threads(int(threads))

    allow_tf32 = precision != "strict"
    torch.backends.cuda.matmul.allow_tf32 = allow_tf32
    torch.backends.cudnn.allow_tf32 = allow_tf32
    torch.backends.cudnn.benchmark = True

    return {"hidden": int(at(first, "128")), "past": int(at(first + 1, "24")),
            "device": device, "precision": precision,
            "autocast": precision == "bf16"}

def autocast_ctx(opts: dict):
    if not opts["autocast"]:
        return contextlib.nullcontext()
    return torch.autocast(device_type=opts["device"], dtype=torch.bfloat16)

def compiled(fn, opts: dict):
    """torch.compile on CUDA, eager on CPU -- both measured, not assumed.

    On CPU, eager is PyTorch's fast path for these models, not a shortcut:
    inductor's CPU codegen loses on a small stack of GEMMs. Measured on this
    machine, dense CPU training, batch 4,096: eager 93,156 samples/s against
    compiled 83,722. The previous suite measured the same thing on different
    hardware (41,523 against 29,449), so it is the codegen and not the box.

    Compiling here anyway would hand OpenNN a win against a PyTorch nobody
    would ship, which is the mirror image of the eager-on-GPU mistake that
    made dense training read 1.29x when it was 1.06x.

    PT_COMPILE_MODE overrides either way, so the choice stays measurable.
    """
    mode = os.environ.get("PT_COMPILE_MODE", "default")
    if mode == "eager" or opts["device"] != "cuda":
        return fn, "eager"
    return torch.compile(fn, mode=None if mode == "default" else mode), f"compile:{mode}"

def batches_of(text: str) -> list[int]:
    return [int(part) for part in text.split(",") if part]

def sync(opts: dict) -> None:
    if opts["device"] == "cuda":
        torch.cuda.synchronize()

def describe(samples: int, features: int, model: nn.Module, opts: dict) -> None:
    """`features` stays in the signature as documentation of the call site;
    it is not reported, because the OpenNN side has no unambiguous
    counterpart and `parameters` pins the model exactly."""
    print(f"samples={samples} past={opts['past']} "
          f"hidden={opts['hidden']} "
          f"parameters={sum(p.numel() for p in model.parameters())}", flush=True)

def train_like(argv: list[str], mode: str) -> int:
    epochs = int(argv[4]) if len(argv) > 4 else 1
    batches = batches_of(argv[5] if len(argv) > 5 else "256")
    opts = parse_opts(argv, 6)

    print(f"baseline_rss_mib={resident_mib():.1f}")
    print(f"engine=pytorch\nmode={mode}\ndevice={opts['device']}")
    report_blas()

    windows, targets = load_series(argv[2], opts["past"])
    x = torch.from_numpy(windows).to(opts["device"])
    y = torch.from_numpy(targets).to(opts["device"])
    warmup = 2 if mode == "train" else 0

    for batch in batches:
        model = build(x.shape[2], opts)
        if batch == batches[0]:
            describe(x.shape[0], x.shape[2], model, opts)

        loss_fn = nn.MSELoss()
        optimizer = torch.optim.Adam(model.parameters())
        ctx = autocast_ctx(opts)

        def step(xb, yb):
            optimizer.zero_grad(set_to_none=True)
            with ctx:
                loss = loss_fn(model(xb), yb)
            loss.backward()
            optimizer.step()

        step_fn, how = compiled(step, opts)
        starts = range(0, x.shape[0] - batch + 1, batch)

        def run_epoch():
            model.train()
            for start in starts:
                step_fn(x[start:start + batch], y[start:start + batch])
            sync(opts)

        for _ in range(warmup):
            run_epoch()

        print(f"TIMED_START_UNIX={time.time():.3f}", flush=True)
        times = []
        for _ in range(epochs):
            mark = time.perf_counter()
            run_epoch()
            times.append(time.perf_counter() - mark)
        print(f"TIMED_END_UNIX={time.time():.3f}", flush=True)

        times.sort()
        median = times[len(times) // 2]

        print(f"batch_{batch}_samples_per_sec={int(len(list(starts)) * batch / median)}"
              f" median_epoch_s={median:.6g}")
        print(f"batch_{batch}_mode={how}", flush=True)

    print("RESULT=OK")
    return 0

def infer(argv: list[str]) -> int:
    reps = int(argv[3]) if len(argv) > 3 else 1
    batches = batches_of(argv[4] if len(argv) > 4 else "256")
    opts = parse_opts(argv, 5)

    print(f"baseline_rss_mib={resident_mib():.1f}")
    print(f"engine=pytorch\nmode=infer\ndevice={opts['device']}")
    report_blas()

    windows, _ = load_series(argv[2], opts["past"])
    x = torch.from_numpy(windows).to(opts["device"])

    for batch in batches:
        model = build(x.shape[2], opts).eval()
        if batch == batches[0]:
            describe(x.shape[0], x.shape[2], model, opts)

        forward, _ = compiled(model, opts)
        processed = (x.shape[0] // batch) * batch
        window = x[:batch]

        def run_pass():
            with torch.no_grad(), autocast_ctx(opts):
                for _ in range(processed // batch):
                    forward(window)
            sync(opts)

        run_pass()

        print(f"TIMED_START_UNIX={time.time():.3f}", flush=True)
        times = []
        for _ in range(reps):
            mark = time.perf_counter()
            run_pass()
            times.append(time.perf_counter() - mark)
        print(f"TIMED_END_UNIX={time.time():.3f}", flush=True)

        times.sort()
        median = times[len(times) // 2]
        print(f"batch_{batch}_samples_per_sec={int(processed / median)}"
              f" median_pass_s={median:.6g}", flush=True)

    print("RESULT=OK")
    return 0

def capacity(argv: list[str]) -> int:
    batch = int(argv[3]) if len(argv) > 3 else 256
    opts = parse_opts(argv, 4)

    print(f"baseline_rss_mib={resident_mib():.1f}")
    print(f"engine=pytorch\nmode=capacity\ndevice={opts['device']}\nbatch={batch}")
    report_blas()

    try:
        windows, targets = load_series(argv[2], opts["past"])
        x = torch.from_numpy(windows[:batch]).to(opts["device"])
        y = torch.from_numpy(targets[:batch]).to(opts["device"])

        model = build(x.shape[2], opts)
        describe(windows.shape[0], x.shape[2], model, opts)

        loss_fn = nn.MSELoss()
        optimizer = torch.optim.Adam(model.parameters())

        with autocast_ctx(opts):
            loss = loss_fn(model(x), y)
        loss.backward()
        optimizer.step()
        sync(opts)
    except torch.cuda.OutOfMemoryError as error:
        print(f"fits=0\nreason={error}\nRESULT=OOM", flush=True)
        return 1
    except RuntimeError as error:
        if "out of memory" not in str(error).lower():
            raise
        print(f"fits=0\nreason={error}\nRESULT=OOM", flush=True)
        return 1

    print("fits=1\nRESULT=OK", flush=True)
    return 0

def main() -> int:
    argv = sys.argv
    mode = argv[1] if len(argv) > 1 else ""

    if mode in ("train", "quality") and len(argv) > 3:
        return train_like(argv, mode)
    if mode == "infer" and len(argv) > 2:
        return infer(argv)
    if mode == "capacity" and len(argv) > 2:
        return capacity(argv)

    print(__doc__, file=sys.stderr)
    return 2

if __name__ == "__main__":
    raise SystemExit(main())

File: benchmarks/families/lstm.cpp
// The LSTM family, defined once, driven four ways.
//
// LSTM forecasting on UCI Beijing PM2.5, hourly, predicting the next
// reading from a window of past ones.
//
//   lstm train    <csv> <csv> [epochs] [batch,...] [hidden] [past] [dev] [prec]
//   lstm infer    <csv>       [reps]   [batch,...] [hidden] [past] [dev] [prec]
//   lstm capacity <csv>       [batch]              [hidden] [past] [dev] [prec]
//
// LSTM rather than the plain recurrent layer, for two reasons. It is the
// architecture sequence-forecasting results are reported on, and both engines
// route it to the *same* NVIDIA kernel -- OpenNN through
// `cudnn_rnn_forward_`, PyTorch through cuDNN behind `nn.LSTM`. That makes
// this the cleanest cell in the matrix: with the arithmetic identical, what is
// left to measure is the surrounding machinery -- data movement, launch
// overhead, the optimiser -- rather than two teams' hand-written kernels.
//
// `set_parameters_pytorch()` initialises to PyTorch's convention, so the two
// engines start from the same distribution rather than merely the same seed.

#include <algorithm>
#include <chrono>
#include <filesystem>
#include <iomanip>
#include <fstream>
#include <iostream>
#include <numeric>
#include <sstream>
#include <string>
#include <vector>

#if defined(__i386__) || defined(__x86_64__) || defined(_M_IX86) || defined(_M_X64)
#include <immintrin.h>
#endif

#ifdef __linux__
#include <unistd.h>
#endif

#include "opennn/core/configuration.h"
#include "opennn/core/tensor_operations.h"
#include "opennn/core/random_utilities.h"
#include "opennn/core/tensor_types.h"
#include "opennn/dataset/time_series_dataset.h"
#include "opennn/neural_network/forward_propagation.h"
#include "opennn/neural_network/layers/dense_layer.h"
#include "opennn/neural_network/layers/long_short_term_memory_layer.h"
#include "opennn/models/models.h"
#include "opennn/training_strategy/adaptive_moment_estimation.h"
#include "opennn/training_strategy/training_strategy.h"

using namespace opennn;
using clock_type = chrono::steady_clock;

namespace
{

double resident_mb()
{
#ifdef __linux__
    // statm reports pages; the second field is resident.
    ifstream statm("/proc/self/statm");
    long total = 0, resident = 0;
    if (statm >> total >> resident)
        return double(resident) * double(sysconf(_SC_PAGESIZE)) / (1024.0 * 1024.0);
#endif
    return 0.0;
}

constexpr Index SEED = 42;

bool enable_flush_denormals()
{
#if defined(__i386__) || defined(__x86_64__) || defined(_M_IX86) || defined(_M_X64)
    // MXCSR bit 15 is flush-to-zero and bit 6 is denormals-are-zero. They are
    // thread-local, so set them before any OpenMP team is created; workers then
    // inherit the same policy. PyTorch's driver calls set_flush_denormal(true).
    constexpr unsigned int ftz_and_daz = (1U << 15U) | (1U << 6U);
    _mm_setcsr(_mm_getcsr() | ftz_and_daz);
    return (_mm_getcsr() & ftz_and_daz) == ftz_and_daz;
#else
    return false;
#endif
}

struct Options
{
    Index hidden = 128;
    Index past = 24;            // one day of hourly readings
    Device device = Device::CUDA;
    Type precision = Type::FP32;
};

unique_ptr<ForecastingLstmNetwork> build(TimeSeriesDataset& dataset, const Options& options)
{
    set_seed(SEED);

    auto network = make_unique<ForecastingLstmNetwork>(dataset.get_shape("Input"),
                                                       Shape{options.hidden},
                                                       dataset.get_shape("Target"));
    network->set_parameters_pytorch();

    return network;
}

unique_ptr<NeuralNetwork> build_inference(const TimeSeriesDataset& dataset,
                                          const Options& options)
{
    set_seed(SEED);

    auto network = make_unique<NeuralNetwork>();
    network->set_task(NetworkTask::Forecasting);

    auto recurrent = make_unique<LongShortTermMemory>(dataset.get_shape("Input"),
                                                       Shape{options.hidden},
                                                       "Tanh", "Sigmoid",
                                                       "long_short_term_memory_layer");
    recurrent->set_return_sequences(false);
    network->add_layer(std::move(recurrent));

    network->add_layer(make_unique<opennn::Dense>(network->get_output_shape(),
                                                  dataset.get_shape("Target"),
                                                  "Identity",
                                                  BatchNormalization::No,
                                                  "forecasting_layer"));

    network->compile();
    network->set_parameters_pytorch();

    return network;
}

unique_ptr<TimeSeriesDataset> open_dataset(const string& path, const Options& options)
{
    cout << "dataset_opened=" << filesystem::absolute(path).string() << "\n" << flush;
    auto dataset = make_unique<TimeSeriesDataset>(path, ",", true, false);
    dataset->set_past_time_steps(options.past);
    dataset->set_future_time_steps(1);

    if (options.device == Device::CUDA)
        dataset->set_storage_mode(Dataset::StorageMode::GPUPersistantData);

    return dataset;
}

Index use_all_valid_windows(TimeSeriesDataset& dataset, SampleRole role,
                            const Options& options)
{
    const Index windows = max(Index(0), dataset.get_samples_number()
                                      - options.past
                                      - dataset.get_future_time_steps() + 1);
    vector<Index> valid(static_cast<size_t>(windows));
    iota(valid.begin(), valid.end(), Index(0));

    // TimeSeriesDataset normally installs a chronological 60/20/20 split.
    // This benchmark, like the PyTorch driver, uses the complete CSV. Keep
    // only starts whose full input window and target are in range; marking
    // every raw row used to append `past` zero-padded pseudo-windows.
    dataset.set_sample_roles(SampleRole::None);
    dataset.set_sample_roles(valid, role);
    return windows;
}

vector<Index> parse_batches(const string& text)
{
    vector<Index> batches;
    stringstream stream(text);

    for (string item; getline(stream, item, ',');)
        if (!item.empty())
            batches.push_back(stoll(item));

    return batches;
}

Options parse_options(int argc, char* argv[], int first)
{
    Options options;

    if (argc > first)     options.hidden = stoll(argv[first]);
    if (argc > first + 1) options.past = stoll(argv[first + 1]);
    if (argc > first + 2) options.device = string(argv[first + 2]) == "cpu" ? Device::CPU : Device::CUDA;
    if (argc > first + 3) options.precision = string(argv[first + 3]) == "bf16" ? Type::BF16 : Type::FP32;

    return options;
}

AdaptiveMomentEstimation* configure(TrainingStrategy& strategy, Index batch)
{
    strategy.set_loss("MeanSquaredError");
    strategy.set_optimization_algorithm("AdaptiveMomentEstimation");

    auto* adam = dynamic_cast<AdaptiveMomentEstimation*>(strategy.get_optimization_algorithm());
    adam->set_batch_size(batch);
    adam->set_display(false);
    adam->set_display_period(1000000);
    adam->set_gradient_clip_norm(0.0f);

    return adam;
}

void describe(const TimeSeriesDataset& dataset, const NeuralNetwork& network, const Options& options)
{
    cout << "samples=" << dataset.get_used_samples_number()
         << " inputs=" << dataset.get_shape("Input").back()
         << " past=" << options.past
         << " hidden=" << options.hidden
         << " parameters=" << network.get_parameters_number() << "\n" << flush;
}

int usage()
{
    cerr << "usage: lstm train    <csv> <csv> [epochs] [batch,...] [hidden] [past] [dev] [prec]\n"
            "       lstm infer    <csv>       [reps]   [batch,...] [hidden] [past] [dev] [prec]\n"
            "       lstm capacity <csv>       [batch]              [hidden] [past] [dev] [prec]\n";
    return 2;
}

}   // namespace

int main(int argc, char* argv[])
{
    cout << "flush_denormals="
         << (enable_flush_denormals() ? "on" : "unsupported") << "\n";

    // Each engine at its best, as PROTOCOL.md requires. The library defaults to
    // Eigen so a plain build behaves like a plain build; a build that has the
    // MKL kernels is told to use them here rather than inheriting them.
    Configuration::instance().set_blas(Blas::Mkl);
    cout << "blas=" << (blas_mkl_available() ? "mkl" : "eigen") << "\n";

    const string mode = argc > 1 ? argv[1] : "";

    if (mode == "train" || mode == "quality")
    {
        if (argc < 4) return usage();

        const Index epochs = argc > 4 ? stoll(argv[4]) : 1;
        const vector<Index> batches = parse_batches(argc > 5 ? argv[5] : "256");
        const Options options = parse_options(argc, argv, 6);

        Configuration::instance().set(options.device, options.precision);
        cout << "baseline_rss_mib=" << resident_mb() << "\n";
        cout << "engine=opennn\nmode=" << mode
             << "\ndevice=" << (options.device == Device::CPU ? "cpu" : "cuda") << "\n";

        auto dataset = open_dataset(argv[2], options);
        const Index samples = use_all_valid_windows(*dataset, SampleRole::Training, options);

        const bool timing = mode == "train";
        const Index warmup = timing ? 2 : 0;

        for (const Index batch : batches)
        {
            auto network = build(*dataset, options);
            if (batch == batches.front()) describe(*dataset, *network, options);

            const bool graph = options.device == Device::CUDA
                               && getenv("OPENNN_NO_CUDA_GRAPH") == nullptr;

            TrainingStrategy strategy(network.get(), dataset.get());
            auto* adam = configure(strategy, batch);
            adam->set_cuda_graph(graph);
            adam->set_maximum_epochs(warmup + epochs);

            const auto unix_now = []
            {
                return chrono::duration<double>(
                    chrono::system_clock::now().time_since_epoch()).count();
            };

            vector<double> epoch_seconds;
            auto previous_mark = clock_type::now();

            adam->post_epoch_callback = [&](Index epoch, float, float, NeuralNetwork*)
            {
                const auto now = clock_type::now();
                const double elapsed = chrono::duration<double>(now - previous_mark).count();
                previous_mark = now;

                if (epoch == warmup - 1)
                    cout << "TIMED_START_UNIX=" << fixed << setprecision(3)
                         << unix_now() << "\n" << defaultfloat;
                else if (epoch >= warmup)
                    epoch_seconds.push_back(elapsed);

                if (epoch == warmup + epochs - 1)
                    cout << "TIMED_END_UNIX=" << fixed << setprecision(3)
                         << unix_now() << "\n" << defaultfloat;
            };

            strategy.train();

            if (Index(epoch_seconds.size()) != epochs)
            {
                cerr << "epoch timing marks missing\n";
                return 1;
            }

            sort(epoch_seconds.begin(), epoch_seconds.end());
            const double median_epoch_s = epoch_seconds[epoch_seconds.size() / 2];

            cout << "batch_" << batch << "_samples_per_sec="
                 << long(double((samples / batch) * batch) / median_epoch_s)
                 << " median_epoch_s=" << median_epoch_s << "\n"
                 << "batch_" << batch << "_cuda_graph="
                 << (!graph ? "off"
                     : adam->get_cuda_graph_capture_failed() ? "failed" : "captured")
                 << "\n" << flush;
        }
    }
    else if (mode == "infer")
    {
        if (argc < 3) return usage();

        const Index reps = argc > 3 ? stoll(argv[3]) : 1;
        const vector<Index> batches = parse_batches(argc > 4 ? argv[4] : "256");
        const Options options = parse_options(argc, argv, 5);

        Configuration::instance().set(options.device, options.precision);
        cout << "baseline_rss_mib=" << resident_mb() << "\n";
        cout << "engine=opennn\nmode=infer\ndevice="
             << (options.device == Device::CPU ? "cpu" : "cuda") << "\n";

        auto dataset = open_dataset(argv[2], options);
        const Index samples = use_all_valid_windows(*dataset, SampleRole::Testing, options);

        for (const Index batch : batches)
        {
            // Match PyTorch's timed nn.LSTM + nn.Linear exactly. The full
            // ForecastingLstmNetwork remains the train/quality/capacity model;
            // its Scaling, Unscaling, and Clamping endpoints are deliberately
            // absent from the inference throughput graph.
            auto network = build_inference(*dataset, options);
            if (batch == batches.front()) describe(*dataset, *network, options);

            const Index processed = (samples / batch) * batch;

            ForwardPropagation forward_propagation(batch, network.get(),
                                                   ForwardPropagationMode::Inference);
            const bool graph = options.device == Device::CUDA
                               && getenv("OPENNN_NO_CUDA_GRAPH") == nullptr;
            forward_propagation.set_cuda_graph(graph);

            vector<Index> indices(size_t(batch), Index(0));
            for (Index k = 0; k < batch; ++k) indices[size_t(k)] = k;

            Batch data(batch, dataset.get(), network->get_config());
            data.fill(indices, dataset->get_feature_selection(), FillMode::Inference);

            // fill() stages a CUDA batch on the host; the transfer is a
            // separate stream operation, issued here once, before the clock,
            // so the replayed pass runs on the windows it was filled with.
            if (options.device == Device::CUDA)
            {
                data.upload_to_device_batch_async(data, device::get_transfer_stream());
                data.wait_h2d_on_compute_stream();
            }

            const vector<TensorView>& inputs = data.get_inputs();

            const auto run_once = [&](bool upload_parameters)
            {
                if (options.device == Device::CUDA)
                    network->calculate_outputs_resident(inputs, forward_propagation,
                                                        upload_parameters);
                else
                    network->forward_propagate(inputs, forward_propagation,
                                               ForwardPropagationMode::Inference);
            };

            const auto run_pass = [&]
            {
                for (Index i = 0; i + batch <= samples; i += batch)
                    run_once(false);

                // CUDA calls are asynchronous. End the pass at completion so
                // this measures kernel execution, as the PyTorch benchmark's
                // torch.cuda.synchronize() does, rather than launch latency.
                if (options.device == Device::CUDA)
                    device::synchronize(device::get_compute_stream());
            };

            // Uploading parameters is a device operation -- a CPU-compiled
            // network has no device copy and `copy_parameters_device` throws,
            // which aborted every CPU inference cell in this family. The
            // warm-up pass still has to happen, so it is the upload that is
            // conditional, not the pass.
            run_once(options.device == Device::CUDA);
            run_pass();

            const auto unix_now = []
            {
                return chrono::duration<double>(
                    chrono::system_clock::now().time_since_epoch()).count();
            };

            cout << "TIMED_START_UNIX=" << fixed << setprecision(3)
                 << unix_now() << "\n" << defaultfloat << flush;

            vector<double> times;
            for (Index r = 0; r < reps; ++r)
            {
                const auto t0 = clock_type::now();
                run_pass();
                times.push_back(chrono::duration<double>(clock_type::now() - t0).count());
            }

            cout << "TIMED_END_UNIX=" << fixed << setprecision(3)
                 << unix_now() << "\n" << defaultfloat << flush;

            sort(times.begin(), times.end());
            const double median_pass_s = times[times.size() / 2];

            cout << "batch_" << batch << "_samples_per_sec="
                 << long(double(processed) / median_pass_s)
                 << " median_pass_s=" << median_pass_s << "\n"
                 << "batch_" << batch << "_cuda_graph="
                 << (!graph ? "off"
                     : forward_propagation.cuda_graph_failed ? "failed"
                     : forward_propagation.inference_graph_exec ? "captured" : "warming")
                 << "\n" << flush;
        }
    }
    else if (mode == "capacity")
    {
        if (argc < 3) return usage();

        const Index batch = argc > 3 ? stoll(argv[3]) : 256;
        const Options options = parse_options(argc, argv, 4);

        Configuration::instance().set(options.device, options.precision);
        cout << "baseline_rss_mib=" << resident_mb() << "\n";
        cout << "engine=opennn\nmode=capacity\ndevice="
             << (options.device == Device::CPU ? "cpu" : "cuda")
             << "\nbatch=" << batch << "\n";

        try
        {
            auto dataset = open_dataset(argv[2], options);
            use_all_valid_windows(*dataset, SampleRole::Training, options);

            auto network = build(*dataset, options);
            describe(*dataset, *network, options);

            TrainingStrategy strategy(network.get(), dataset.get());
            configure(strategy, batch)->set_maximum_epochs(1);
            strategy.train();
        }
        catch (const exception& error)
        {
            cout << "fits=0\nreason=" << error.what() << "\nRESULT=OOM\n" << flush;
            return 1;
        }

        cout << "fits=1\nRESULT=OK\n" << flush;
        return 0;
    }
    else
    {
        return usage();
    }

    cout << "RESULT=OK\n";

    return 0;
}
